### PR TITLE
Support account locking for TOTP

### PR DESCRIPTION
**Pull request by AnuradhaSK**
## Purpose
Support account locking for TOTP

Similar behaviour in wso2-extensions/identity-outbound-auth-sms-otp#91 and wso2-extensions/identity-outbound-auth-email-otp#79

### CODE REVIEW

File: component/authenticator/src/main/java/org/wso2/carbon/identity/application/authenticator/totp/util/TOTPUtil.java [hunk 1]
```diff
@@ -555,4 +557,76 @@ public static boolean isEventHandlerBasedEmailSenderEnabled() {
...
+	 *
+	 * @return True if account locking is enabled for TOTP.
+	 */
+	public static boolean isAccountLockingEnabledForTotp() {
+
+		return Boolean.parseBoolean(
+				getTOTPParameters().get(TOTPAuthenticatorConstants.ENABLE_ACCOUNT_LOCKING_FOR_FAILED_ATTEMPTS));
+	}
+
+	public static boolean isAccountLocked(String userName, String tenantDomain, String userStoreDomain)
```
Comment (sachiniWettasinghe) [inline component/authenticator/src/main/java/org/wso2/carbon/identity/application/authenticator/totp/util/TOTPUtil.java:599]: Add method description

File: component/authenticator/src/main/java/org/wso2/carbon/identity/application/authenticator/totp/util/TOTPUtil.java [hunk 2]
```diff
@@ -555,4 +557,76 @@ public static boolean isEventHandlerBasedEmailSenderEnabled() {
...
+			throws AuthenticationFailedException {
+
+		try {
+			return TOTPDataHolder.getInstance().getAccountLockService()
+					.isAccountLocked(userName, tenantDomain, userStoreDomain);
+		} catch (Exception e) {
```
Comment (sachiniWettasinghe) [inline component/authenticator/src/main/java/org/wso2/carbon/identity/application/authenticator/totp/util/TOTPUtil.java:605]: Is there any specific reason here to catch Exception and throw a subclass of Exception?

File: component/authenticator/src/main/java/org/wso2/carbon/identity/application/authenticator/totp/util/TOTPUtil.java [hunk 3]
```diff
@@ -555,4 +557,76 @@ public static boolean isEventHandlerBasedEmailSenderEnabled() {
...
+			connectorConfigs = TOTPDataHolder.getInstance()
+					.getIdentityGovernanceService()
+					.getConfiguration(
+							new String[]{
+									TOTPAuthenticatorConstants.PROPERTY_ACCOUNT_LOCK_ON_FAILURE,
+									TOTPAuthenticatorConstants.PROPERTY_ACCOUNT_LOCK_ON_FAILURE_MAX,
+									TOTPAuthenticatorConstants.PROPERTY_ACCOUNT_LOCK_TIME,
+									TOTPAuthenticatorConstants.PROPERTY_LOGIN_FAIL_TIMEOUT_RATIO
+							}, tenantDomain);
+		} catch (Exception e) {
```
Comment (sachiniWettasinghe) [inline component/authenticator/src/main/java/org/wso2/carbon/identity/application/authenticator/totp/util/TOTPUtil.java:581]: Is there any specific reason here to catch Exception and throw a subclass of Exception?

File: component/authenticator/src/main/java/org/wso2/carbon/identity/application/authenticator/totp/util/TOTPUtil.java [hunk 4]
```diff
@@ -555,4 +557,76 @@ public static boolean isEventHandlerBasedEmailSenderEnabled() {
...
+	 * @param context Authentication context.
+	 * @return Whether the user being authenticated via a local authenticator.
+	 */
+	public static boolean isLocalUser(AuthenticationContext context) {
+
+		Map<Integer, StepConfig> stepConfigMap = context.getSequenceConfig().getStepMap();
+		if(stepConfigMap != null) {
+			for (StepConfig stepConfig : stepConfigMap.values()) {
+				if (stepConfig.getAuthenticatedUser() != null && stepConfig.isSubjectAttributeStep()) {
+					if (stepConfig.getAuthenticatedIdP().equals(TOTPAuthenticatorConstants.LOCAL_AUTHENTICATOR)) {
```
Comment (sachiniWettasinghe) [inline component/authenticator/src/main/java/org/wso2/carbon/identity/application/authenticator/totp/util/TOTPUtil.java:623]: Shall we improve the logic by merging these two if conditions?

File: component/authenticator/src/main/java/org/wso2/carbon/identity/application/authenticator/totp/util/TOTPUtil.java (hunk 3 above)
Comment (AnuradhaSK) [reply component/authenticator/src/main/java/org/wso2/carbon/identity/application/authenticator/totp/util/TOTPUtil.java:581]: `processAuthenticationResponse` method only throws `AuthenticationFailedException` to handle in `AbstractApplicationAuthenticator`. This method is invoked inside `processAuthenticationResponse`

File: component/authenticator/src/main/java/org/wso2/carbon/identity/application/authenticator/totp/util/TOTPUtil.java (hunk 2 above)
Comment (AnuradhaSK) [reply component/authenticator/src/main/java/org/wso2/carbon/identity/application/authenticator/totp/util/TOTPUtil.java:614]: `processAuthenticationResponse` method only throws `AuthenticationFailedException` to handle in `AbstractApplicationAuthenticator`. This method is invoked inside `processAuthenticationResponse`

File: component/authenticator/src/main/java/org/wso2/carbon/identity/application/authenticator/totp/util/TOTPUtil.java (hunk 1 above)
Comment (AnuradhaSK) [reply component/authenticator/src/main/java/org/wso2/carbon/identity/application/authenticator/totp/util/TOTPUtil.java:608]: Fixed in d19089040144760714d21b7650d1bc8785003028

File: component/authenticator/src/main/java/org/wso2/carbon/identity/application/authenticator/totp/util/TOTPUtil.java (hunk 4 above)
Comment (AnuradhaSK) [reply component/authenticator/src/main/java/org/wso2/carbon/identity/application/authenticator/totp/util/TOTPUtil.java]: Fixed in d19089040144760714d21b7650d1bc8785003028

File: component/authenticator/src/main/java/org/wso2/carbon/identity/application/authenticator/totp/util/TOTPUtil.java (hunk 3 above)
Comment (sachiniWettasinghe) [reply component/authenticator/src/main/java/org/wso2/carbon/identity/application/authenticator/totp/util/TOTPUtil.java:581]: What I meant here is, we should catch the exact exception type which can be thrown from the try block. Is there any specific reason for not to do so?

Comment (AnuradhaSK) [reply component/authenticator/src/main/java/org/wso2/carbon/identity/application/authenticator/totp/util/TOTPUtil.java]: Fixed https://github.com/wso2-extensions/identity-outbound-auth-totp/pull/73/commits/7855d43085638b13d6de92a0460ba4267a20bb9e

File: component/authenticator/src/main/java/org/wso2/carbon/identity/application/authenticator/totp/util/TOTPUtil.java (hunk 2 above)
Comment (AnuradhaSK) [reply component/authenticator/src/main/java/org/wso2/carbon/identity/application/authenticator/totp/util/TOTPUtil.java]: Fixed  https://github.com/wso2-extensions/identity-outbound-auth-totp/pull/73/commits/7855d43085638b13d6de92a0460ba4267a20bb9e

**Pull request merged**